torch/_refs/__init__.py

```diff
@@ -0,0 +1,588 @@
+    prims.bitwise_xor, type_promotion=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT
+)
+
+# TODO: add docstring
+# complex =  _make_elementwise_binary_reference(prims.complex, type_promotion=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT)
+
+# TODO: add docstring
+eq = _make_elementwise_binary_reference(
+    prims.eq, type_promotion=ELEMENTWISE_TYPE_PROMOTION_KIND.ALWAYS_BOOL
+)
+
+# TODO: add docstring
+# Float power has its own implementation because it has unique type promotion.
+def float_power(
+    a: Union[Tensor, Number], b: Union[Tensor, Number], out: Optional[Tensor] = None
+) -> Tensor:
+
+    assert isinstance(a, (Tensor, Number))
+    assert isinstance(b, (Tensor, Number))
+    assert out is None or isinstance(out, TensorLike)
+
+    # Special-cases Number x Number case
+    if isinstance(a, Number) and isinstance(b, Number):
+        a, b = utils.wrap_scalars(a, b)
+
+    # Handles type promotion
+    dtype = utils.get_higher_dtype(a, b)
+    if utils.is_complex_dtype(dtype):
+        dtype = torch.complex128
+    else:
+        dtype = torch.float64
+
+    a, b = _convert_dtype(a, b, dtype=dtype)
+
+    # Broadcasting
+    a, b = broadcast(a, b)
+
+    result = prims.pow(a, b)
+
+    if out is not None:
+        out = _maybe_resize_out(out, result.shape)
+        return copy_to(out, result, allow_cross_device=False)
+
+    return result
+
+
+# TODO: add docstring
+ge = _make_elementwise_binary_reference(
+    prims.ge, type_promotion=ELEMENTWISE_TYPE_PROMOTION_KIND.ALWAYS_BOOL
+)
+
+# TODO: add docstring
+gt = _make_elementwise_binary_reference(
+    prims.gt, type_promotion=ELEMENTWISE_TYPE_PROMOTION_KIND.ALWAYS_BOOL
+)
+
+# TODO: add docstring
+le = _make_elementwise_binary_reference(
+    prims.lt, type_promotion=ELEMENTWISE_TYPE_PROMOTION_KIND.ALWAYS_BOOL
+)
+
+# TODO: add docstring
+lt = _make_elementwise_binary_reference(
+    prims.lt, type_promotion=ELEMENTWISE_TYPE_PROMOTION_KIND.ALWAYS_BOOL
+)
+
+# TODO: add docstring
+maximum = _make_elementwise_binary_reference(
+    prims.max, type_promotion=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT
+)
+
+# TODO: add docstring
+minimum = _make_elementwise_binary_reference(
+    prims.min, type_promotion=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT
+)
+
+# TODO: add docstring
+mul = _make_elementwise_binary_reference(
+    prims.mul, type_promotion=ELEMENTWISE_TYPE_PROMOTION_KIND.OP_MATH
+)
+
+# TODO: add docstring
+ne = _make_elementwise_binary_reference(
+    prims.ne, type_promotion=ELEMENTWISE_TYPE_PROMOTION_KIND.ALWAYS_BOOL
+)
+
+# TODO: add docstring
+nextafter = _make_elementwise_binary_reference(
+    prims.nextafter, type_promotion=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT
+)
+
+# TODO: add docstring
+pow = _make_elementwise_binary_reference(
+    prims.pow, type_promotion=ELEMENTWISE_TYPE_PROMOTION_KIND.OP_MATH
+)
+
+# TODO: add docstring
+# Sub is implemented specially because it has an alpha argument and
+#   is decomposed into multiple prims.
+# TODO: consider refactoring this with add impl
```

Comment:
in PyTorch sub(a, b, alpha) is defined as add(a, b, -alpha) and it seems like this should be possible here

Reply:
That's true but the benefit of doing so is ???

Modeling a primitive sub operation should be trivial for trace executors

Reply:
Less code to write? The ref for add and sub are pretty complicated because they need to be multiple prim ops. Seems worth cutting it down

Reply:
Sure